tools/swift-inspect/Sources/swift-inspect/LinuxRemoteProcess.swift

```diff
@@ -0,0 +1,155 @@
+    public private(set) var processName: String = "<unknown process>"
+
+    let memoryMap: SwiftInspectLinux.MemoryMap
+    let symbolCache: SwiftInspectLinux.SymbolCache
+
+    static var QueryDataLayout: QueryDataLayoutFunction {
+      return { (context, type, _, output) in
+        guard let output = output else { return 0 }
+
+        switch type {
+        case DLQ_GetPointerSize:
+          let size = UInt8(MemoryLayout<UnsafeRawPointer>.stride)
+          output.storeBytes(of: size, toByteOffset: 0, as: UInt8.self)
+          return 1
+
+        case DLQ_GetSizeSize:
+          let size = UInt8(MemoryLayout<UInt>.stride)  // UInt is word-size like size_t
+          output.storeBytes(of: size, toByteOffset: 0, as: UInt8.self)
+          return 1
+
+        case DLQ_GetLeastValidPointerValue:
+          let value: UInt64 = 0x1000
+          output.storeBytes(of: value, toByteOffset: 0, as: UInt64.self)
+          return 1
+
+        default: return 0
+        }
+      }
+    }
+
+    static var Free: FreeFunction {
+      return { (_, bytes, _) in free(UnsafeMutableRawPointer(mutating: bytes)) }
+    }
+
+    static var ReadBytes: ReadBytesFunction {
+      return { (context, address, size, _) in
+        let process: LinuxRemoteProcess = LinuxRemoteProcess.fromOpaque(context!)
+
+        guard
+          let byteArray: [UInt8] = try? process.process.readArray(
+            address: address, upToCount: UInt(size)), let buffer = malloc(byteArray.count)
+        else { return nil }
+
+        byteArray.withUnsafeBytes {
+          buffer.copyMemory(from: $0.baseAddress!, byteCount: byteArray.count)
+        }
+
+        return UnsafeRawPointer(buffer)
+      }
+    }
+
+    static var GetStringLength: GetStringLengthFunction {
+      return { (context, address) in
+        let process: LinuxRemoteProcess = LinuxRemoteProcess.fromOpaque(context!)
+
+        // copy the string from the remote proces to get its length
+        guard let string = try? process.process.readString(address: address),
+          let len = UInt64(exactly: string.count)
+        else { return 0 }
+        return len
+      }
+    }
+
+    static var GetSymbolAddress: GetSymbolAddressFunction {
+      return { (context, symbol, length) in
+        let process: LinuxRemoteProcess = LinuxRemoteProcess.fromOpaque(context!)
+
+        guard let symbol = symbol else { return 0 }
+        let name: String = symbol.withMemoryRebound(to: UInt8.self, capacity: Int(length)) {
+          let buffer = UnsafeBufferPointer(start: $0, count: Int(length))
+          return String(decoding: buffer, as: UTF8.self)
+        }
+
+        guard let (startAddr, _) = process.symbolCache.address(of: name) else { return 0 }
+        return startAddr
+      }
+    }
+
+    init?(processId: ProcessIdentifier) {
+      self.processIdentifier = processId
+      do {
+        let path = "/proc/\(processId)/cmdline"
+        self.processName = try String(contentsOfFile: path, encoding: .ascii)
```

Comment:
Should this be `.utf8`? I don't think any encoding beyond bag-of-bytes is entirely correct, but UTF-8 would stand a chance of doing the right thing for a non-ASCII filename.